### ISSUE TITLE
Bug: `tsconfig.json` not respected in new files when multiple tsconfigs include the same directory with `useProjectService` enabled

### ISSUE THREAD

**Issue by fire332** [typescript-eslint/typescript-eslint#8835]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Issue Description

When `EXPERIMENTAL_useProjectService: true` and at least two `tsconfig.json`s include a non-empty common directory, typescript-eslint will not respect `tsconfig.json` for new/renamed files until ESLint is reloaded.

Possibly related: #8206 #7435

### Reproduction Repository Link

https://github.com/fire332/typescript-eslint-bug-repro/

### Repro Steps

1) clone the repo
2) `pnpm install`
3) open project in VSCode with the ESLint extension installed
4) **IMPORTANT**: open `.prettierrc.js` or `eslint.config.js` and wait for ESLint to initialize
5) create a new `.ts` file in `./src` or `./src-alt`
6) problems reported will show:
```log
This rule requires the `strictNullChecks` compiler option to be turned on to function correctly. eslint(@typescript-eslint/no-unnecessary-condition)
```
7) run `Developer: Reload Window` from the command palette
8) check the file you created and note there's no problems reported
9) rename the file and note that the same problem will be reported again

### Versions

| package                                | version |
| -------------------------------------- | ------- |
| `typescript-eslint`     | `7.5.0` |
| `TypeScript`                           | `5.4.3` |
| `ESLint`                               | `8.57.0` |
| `node`                                 | `>=20.11.1` |

**Comment by karlhorky** [comment]
I can confirm this behavior of erroring with the `strictNullChecks` error message, although I'm not sure about the requirement that there are multiple `tsconfig.json` files - in my reproduction, it also happens with a single `tsconfig.json` file 🤔

My reproduction:

1. Open a `.tsx` file (confirm that linting is working)
2. Copy the file
3. 💥 Observe the error


https://github.com/typescript-eslint/typescript-eslint/assets/1935696/67af34fe-fc48-4af8-b929-726560ce80a1

Error message:

```
This rule requires the `strictNullChecks` compiler option to be turned on to function correctly.eslint@typescript-eslint/no-unnecessary-condition
```

Reproduction files:

`eslint.config.js`

```js
import eslintTypescript from '@typescript-eslint/eslint-plugin';
import typescriptParser from '@typescript-eslint/parser';

/** @type {import('@typescript-eslint/utils/ts-eslint').FlatConfig.ConfigArray} */
const config = [
  {
    files: ['**/*.js', '**/*.jsx', '**/*.ts', '**/*.tsx'],
    languageOptions: {
      parser: typescriptParser,
      parserOptions: {
        project: './tsconfig.json',
        EXPERIMENTAL_useProjectService: true,
        warnOnUnsupportedTypeScriptVersion: true,
      },
    },
    plugins: {
      '@typescript-eslint': {
        rules: eslintTypescript.rules,
      },
    },
    rules: {
      '@typescript-eslint/no-unnecessary-condition': 'warn',
    },
  },
];

export default config;
```

`tsconfig.json`

```json
{
  "$schema": "https://json.schemastore.org/tsconfig",
  "compilerOptions": {
    "lib": ["DOM", "DOM.Iterable", "ESNext"],
    "target": "ES2015",
    "moduleResolution": "Bundler",
    "moduleDetection": "force",
    "resolveJsonModule": true,
    "esModuleInterop": true,
    "isolatedModules": true,
    "allowJs": true,
    "allowSyntheticDefaultImports": true,
    "downlevelIteration": true,
    "forceConsistentCasingInFileNames": true,
    "noEmit": true,
    "noFallthroughCasesInSwitch": true,
    "skipLibCheck": true,
    "strict": true,
    "incremental": true,
    "noUncheckedIndexedAccess": true,
    "checkJs": true,
    "jsx": "preserve",
    "strictNullChecks": true,
    "module": "esnext"
  },
  "include": [
    "**/*.ts",
    "**/*.tsx",
    "**/*.js",
    "**/*.jsx",
    "**/*.cjs",
    "**/*.mjs"
  ],
  "exclude": ["node_modules", "build", "public", "scripts"]
}
```

**Comment by JoshuaKGoldberg** [comment]
Hey good news, I tried this out again today and it looks like the problem is fixed! I tried the two mentioned reproductions and a new clean repo going off our _Getting Started_ docs.

I haven't dug deep but I my hunch is some combination of #8815 + #9097 would at least be good places to look.

@fire332 @karlhorky are you able to confirm whether this issue is still happening for you? 🙏

**Comment by fire332** [comment]
Without changing any of the configs and just updating dependencies, the behavior described in the opening comment remains identical.

@JoshuaKGoldberg Have you ensured the ESLint extension is installed and ESLint completed initializing in step 4 before moving on to step 5?

### Versions


| package                                | version |
| -------------------------------------- | ------- |
| `typescript-eslint`     | **NEW** `7.11.0` |
| `TypeScript`                           | **NEW** `5.4.5` |
| `ESLint`                               | `8.57.0` |
| `node`                                 | **NEW** `20.14.0` |

**Comment by karlhorky** [comment]
I'm also still experiencing [the problem described in my comment](https://github.com/typescript-eslint/typescript-eslint/issues/8835#issuecomment-2041136376), using ESLint v9, the latest v8 alpha and [the new `languageOptions.parserOptions.projectService` config option naming](https://github.com/typescript-eslint/typescript-eslint/pull/9084).

- `typescript-eslint@8.0.0-alpha.16`
- `eslint@9.3.0`
- `typescript@5.4.5`

**Comment by JoshuaKGoldberg** [comment]
Hmm. Interesting. I'll take another look, thanks.

**Comment by JoshuaKGoldberg** [comment]
Hmm. Interesting. I'll take another look, thanks.

In case it's useful, I'm only testing with v8 - as of yesterday, `typescript-eslint@8.0.0-alpha.25`. We have some improvements to the project service on the v8 line that we don't plan on backporting to v7.

**Comment by fire332** [comment]
I've just tested on the v8 version with ESLint v9 with the patch from #9205 as well and the same issue occurs. The repro has been updated on the `typescript-eslint@8` branch.

**Comment by higherorderfunctor** [comment]
@fire332 my patch would fix your base compiler options not getting pulled in where you extends in the [tsconfig.json](https://github.com/fire332/typescript-eslint-bug-repro/blob/229820a299db670c2faa1773abd780ff0d1b31d3/tsconfig.json#L2).

Stuff like this should start working.

```json
"paths": {
  "@common/*": ["./common/*"],
},
```

Two thoughts:

First, are we sure this isn't a tsserver issue?  To find the bug for my patch I had to enable logging here: https://github.com/typescript-eslint/typescript-eslint/blob/b4fbf6dda732d28f54531d6aa19e8ecfb0e47c9f/packages/typescript-estree/src/create-program/createProjectService.ts#L56

With something like this:

```ts
        logger: {
            close: doNothing,
            endGroup: doNothing,
            getLogFileName: () => undefined,
            hasLevel: () => true,
            info: (s) => console.info(`INFO: ${s}`),
            loggingEnabled: () => true,
            msg: (s, type) => console.log(`MSG (${type}): ${s}`),
            perftrc: (s) => console.log(`PERF: ${s}`),
            startGroup: doNothing,
        },
```

I was using the CLI, so if this is an editor only bug you may need to write to a file unless your editor can capture stdout.

You can also try the [environment variable](https://github.com/microsoft/TypeScript/wiki/Standalone-Server-(tsserver)#logging), but it didn't work for me on the CLI running eslint.  You may also get double logs if your editor runs tsserver + eslint language servers that could overwrite themselves so probably best to leave the filename alone to append the PID if this happens to work through vscode.

I also ran with the environment variable `DEBUG=*` to log all eslint messages.

With logging in eslint and tsserver I was able to trace the issue I found.  I raise that it _might_ be tsserver as I see file watchers in my logs like this:

```
INFO: FileWatcher:: Added:: WatchInfo: /<PROJECT>/tsconfig.json 2000 undefined Project: /<PROJECT>/tsconfig.json WatchType: Config file
```

@JoshuaKGoldberg would you accept a PR to allow the tsserver logging to be enabled?  It was really helpful in running down project service related issues.

The second thought is why are we overlapping includes in tsconfigs?

The setup I use is below.  All `tsconfig.json`s are simply "linchpins" with an explicit "include: []" and some project references.  This is how my tooling finds all it's files from the root `tsconfig.json`.  `tsconfig.build.json` has overlapping includes but the root `tsconfig.build.json` has its own graph never seen by my editor.

A large scale project that follows this method is https://github.com/Effect-TS/effect if you wish to look at an extended example.

```sh
.
├── packages
│   ├── package-a
│   │   ├── tsconfig.build.json   # includes lib (or src) files, has refs to workspace dependencies' tsconfig.build.json
│   │   ├── tsconfig.json         # has refs to lib and test
│   │   └── tsconfig.lib.json     # includes lib (or src) files, has refs to workspace dependencies' tsconfig.json
│   │   └── tsconfig.test.json    # includes tests files, has refs to lib (or src) and workspace dependencies' tsconfig.json
│   ├── package-b
│   │   ├── tsconfig.build.json   # includes lib (or src) files,has refs to workspace dependencies' tsconfig.build.json
│   │   ├── tsconfig.json         # has refs to lib and test
│   │   └── tsconfig.lib.json     # includes lib (or src) files, has refs to workspace dependencies' tsconfig.json
│   │   └── tsconfig.test.json    # includes tests files, has refs to lib (or src) and workspace dependencies' tsconfig.json
├── tsconfig.base.json            # has shared compiler options, EVERY other tsconfig extends this one
├── tsconfig.build.json           # has refs to each package's tsconfig.build.json
├── tsconfig.json                 # has refs to each package's tsconfig.json and tsconfig.utils.json
├── tsconfig.utils.json           # includes root level stuff like my eslint.config.js, esbuild.config.js, etc
```

For a large project you will need to increase `maximumDefaultProjectFileMatchCount_THIS_WILL_SLOW_DOWN_LINTING` from https://typescript-eslint.io/packages/typescript-estree/.

Keep in mind that this is pretty slow as the option indicates.  My linting for my work project has gone from a couple minutes to 20 minutes and I suspect there is some race condition as it sometimes goes for an hour before the CI server kills it.  Most of that overhead appears to just be tsserver resolving files.  My project is a mixed frontend/backend monorepo using vue which doesn't help since it looks, at a glance at least so take with a grain of salt, to send every code block in a .vue file to be resolved (e.g., I think `<h3>{{ section.text }}</h3>` is maybe sent as its own file).  Excluding vue files it is roughly 3x slower (2min to 6min on my project).  

Here is an example of a vue file that seems to keep getting opened on a single file lint from the cli that shows the issue.

```
  typescript-eslint:typescript-estree:useProgramFromProjectService Opening project service file for: <REDACTED>/project/packages/project-a/src/views/Home.vue at absolute path <REDACTED>/project/packages/project-a/src/views/Home.vue +0ms
  typescript-eslint:typescript-estree:useProgramFromProjectService Opened project service file: { configFileName: '<REDACTED>/project/packages/project-a/tsconfig.src.json', configFileErrors: [] } +5s
  typescript-eslint:typescript-estree:useProgramFromProjectService Retrieving script info and then program for: <REDACTED>/project/packages/project-a/src/views/Home.vue +1ms
  typescript-eslint:typescript-estree:useProgramFromProjectService Found project service program for: <REDACTED>/project/packages/project-a/src/views/Home.vue +0ms
  typescript-eslint:typescript-estree:useProgramFromProjectService Opening project service file for: <REDACTED>/project/packages/project-a/src/views/Home.vue at absolute path <REDACTED>/project/packages/project-a/src/views/Home.vue +8ms
  typescript-eslint:typescript-estree:useProgramFromProjectService Opened project service file: { configFileName: '<REDACTED>/project/packages/project-a/tsconfig.src.json', configFileErrors: [] } +1s
  typescript-eslint:typescript-estree:useProgramFromProjectService Retrieving script info and then program for: <REDACTED>/project/packages/project-a/src/views/Home.vue +0ms
  typescript-eslint:typescript-estree:useProgramFromProjectService Found project service program for: <REDACTED>/project/packages/project-a/src/views/Home.vue +0ms
  typescript-eslint:typescript-estree:useProgramFromProjectService Opening project service file for: <REDACTED>/project/packages/project-a/src/views/Home.vue at absolute path <REDACTED>/project/packages/project-a/src/views/Home.vue +2ms
  typescript-eslint:typescript-estree:useProgramFromProjectService Opened project service file: { configFileName: '<REDACTED>/project/packages/project-a/tsconfig.src.json', configFileErrors: [] } +1s
  typescript-eslint:typescript-estree:useProgramFromProjectService Retrieving script info and then program for: <REDACTED>/project/packages/project-a/src/views/Home.vue +1ms
  typescript-eslint:typescript-estree:useProgramFromProjectService Found project service program for: <REDACTED>/project/packages/project-a/src/views/Home.vue +0ms
  typescript-eslint:typescript-estree:useProgramFromProjectService Opening project service file for: <REDACTED>/project/packages/project-a/src/views/Home.vue at absolute path <REDACTED>/project/packages/project-a/src/views/Home.vue +1ms
  typescript-eslint:typescript-estree:useProgramFromProjectService Opened project service file: { configFileName: '<REDACTED>/project/packages/project-a/tsconfig.src.json', configFileErrors: [] } +1s
  typescript-eslint:typescript-estree:useProgramFromProjectService Retrieving script info and then program for: <REDACTED>/project/packages/project-a/src/views/Home.vue +0ms
  typescript-eslint:typescript-estree:useProgramFromProjectService Found project service program for: <REDACTED>/project/packages/project-a/src/views/Home.vue +0ms
  typescript-eslint:typescript-estree:useProgramFromProjectService Opening project service file for: <REDACTED>/project/packages/project-a/src/views/Home.vue at absolute path <REDACTED>/project/packages/project-a/src/views/Home.vue +1ms
  typescript-eslint:typescript-estree:useProgramFromProjectService Opened project service file: { configFileName: '<REDACTED>/project/packages/project-a/tsconfig.src.json', configFileErrors: [] } +1s
  typescript-eslint:typescript-estree:useProgramFromProjectService Retrieving script info and then program for: <REDACTED>/project/packages/project-a/src/views/Home.vue +0ms
  typescript-eslint:typescript-estree:useProgramFromProjectService Found project service program for: <REDACTED>/project/packages/project-a/src/views/Home.vue +0ms
  typescript-eslint:typescript-estree:useProgramFromProjectService Opening project service file for: <REDACTED>/project/packages/project-a/src/views/Home.vue at absolute path <REDACTED>/project/packages/project-a/src/views/Home.vue +1ms
  typescript-eslint:typescript-estree:useProgramFromProjectService Opened project service file: { configFileName: '<REDACTED>/project/packages/project-a/tsconfig.src.json', configFileErrors: [] } +1s
  typescript-eslint:typescript-estree:useProgramFromProjectService Retrieving script info and then program for: <REDACTED>/project/packages/project-a/src/views/Home.vue +0ms
  typescript-eslint:typescript-estree:useProgramFromProjectService Found project service program for: <REDACTED>/project/packages/project-a/src/views/Home.vue +0ms
  typescript-eslint:typescript-estree:useProgramFromProjectService Opening project service file for: <REDACTED>/project/packages/project-a/src/views/Home.vue at absolute path <REDACTED>/project/packages/project-a/src/views/Home.vue +1ms
  typescript-eslint:typescript-estree:useProgramFromProjectService Opened project service file: { configFileName: '<REDACTED>/project/packages/project-a/tsconfig.src.json', configFileErrors: [] } +1s
  typescript-eslint:typescript-estree:useProgramFromProjectService Retrieving script info and then program for: <REDACTED>/project/packages/project-a/src/views/Home.vue +0ms
  typescript-eslint:typescript-estree:useProgramFromProjectService Found project service program for: <REDACTED>/project/packages/project-a/src/views/Home.vue +0ms
  typescript-eslint:typescript-estree:useProgramFromProjectService Opening project service file for: <REDACTED>/project/packages/project-a/src/views/Home.vue at absolute path <REDACTED>/project/packages/project-a/src/views/Home.vue +1ms
  typescript-eslint:typescript-estree:useProgramFromProjectService Opened project service file: { configFileName: '<REDACTED>/project/packages/project-a/tsconfig.src.json', configFileErrors: [] } +1s
  typescript-eslint:typescript-estree:useProgramFromProjectService Retrieving script info and then program for: <REDACTED>/project/packages/project-a/src/views/Home.vue +0ms
  typescript-eslint:typescript-estree:useProgramFromProjectService Found project service program for: <REDACTED>/project/packages/project-a/src/views/Home.vue +0ms
```

**Comment by JoshuaKGoldberg** [comment]
> @JoshuaKGoldberg would you accept a PR to allow the tsserver logging to be enabled? It was really helpful in running down project service related issues.

Absolutely, that'd be great. Yes please! I'd consider it a different issue from this: feel free to file a new issue and/or send a PR. If you do only one, I can do the other. 🙂

**Comment by higherorderfunctor** [comment]
Did some log analysis to see if I could find the inflection point between the original file and a copied file as I am able to reproduce on a large repo.  It did not work on a a small example repo.  That line is where the difference occurs.

**Original File**

```ts
typescript-eslint:typescript-estree:useProgramFromProjectService Opening project service file for: <REDACTED>/packages/project-a/src/stores/auth.ts at absolute path <REDACTED>/packages/project-a/src/stores/auth.ts
typescript-eslint:tsserver INFO: 'Search path: <REDACTED>/packages/project-a/src/stores'
typescript-eslint:tsserver INFO: 'For info: <REDACTED>/packages/project-a/src/stores/auth.ts :: Config file name: <REDACTED>/packages/project-a/tsconfig.json'
typescript-eslint:tsserver INFO: 'Creating configuration project <REDACTED>/packages/project-a/tsconfig.json'
```

That log message happens here: https://github.com/microsoft/TypeScript/blob/27bcd4cb5a98bce46c9cdd749752703ead021a4b/src/server/editorServices.ts#L2462

**Copied File**

```ts
typescript-eslint:typescript-estree:useProgramFromProjectService Opening project service file for: <REDACTED>/packages/project-a/src/stores/auth2.ts at absolute path <REDACTED>/packages/project-a/src/stores/auth2.ts
typescript-eslint:tsserver INFO: 'Search path: <REDACTED>/packages/project-a/src/stores'
typescript-eslint:tsserver INFO: 'For info: <REDACTED>/packages/project-a/src/stores/auth2.ts :: Config file name: <REDACTED>/packages/project-a/tsconfig.json'
typescript-eslint:tsserver INFO: 'FileWatcher:: Added:: WatchInfo: <REDACTED>/packages/project-a/src/stores/tsconfig.json 2000 undefined WatchType: Config file for the inferred project root'
```

This log message happens here: https://github.com/microsoft/TypeScript/blob/27bcd4cb5a98bce46c9cdd749752703ead021a4b/src/server/editorServices.ts#L2192

I haven't had the time to do a full code level trace or fire up the debugger yet, but at a glance it looks like tsserver treats the new file as an orphan.  Since I cannot reproduce on a small repo, it may suggest a race condition.

Wrapping up for the night I took a shot in the dark at changing this line in the plugin to use `getScriptInfoEnsuringProjectsUptoDate` instead of `getScriptInfo`: https://github.com/typescript-eslint/typescript-eslint/blob/c9a6dd9f3df4e3c2bfcc5dceb770b021709152ae/packages/typescript-estree/src/useProgramFromProjectService.ts#L73

It didn't fix the issue, but this week is also my first venture into the tsserver code, so I may not be using it correctly.

**Comment by fluidsonic** [comment]
I ran into the same warning but with a different setup.
>This rule requires the `strictNullChecks` compiler option to be turned on to function correctly

I don't use `EXPERIMENTAL_useProjectService: true` but there may be a common root cause.
If I use `--fix` I get the warning, if I don't use `--fix` I don't get that warning.

I did some debugging. It looks like `typescript-estree` creates two different "programs" for the same file internally, because `--fix` uses multiple cycles:

https://github.com/typescript-eslint/typescript-eslint/blob/63e53e28f68f84764ec632d502da3d33095f6a0d/packages/typescript-estree/src/parser.ts#L235-L253

The first program is created through `getProgramAndAST`, the second one using `createIsolatedProgram`.

The first cycle uses `getProgramAndAST` which in turn uses `createProgramFromConfigFile` to load the correct `tsconfig.json` and pass the options to `createProgram`:
https://github.com/typescript-eslint/typescript-eslint/blob/63e53e28f68f84764ec632d502da3d33095f6a0d/packages/typescript-estree/src/create-program/useProvidedPrograms.ts#L55-L89

The second cycle uses `createIsolatedProgram` which completely ignores `tsconfig.json` and uses pre-defined options:
https://github.com/typescript-eslint/typescript-eslint/blob/63e53e28f68f84764ec632d502da3d33095f6a0d/packages/typescript-estree/src/create-program/createIsolatedProgram.ts#L64-L74

That in turn explains the warning because `strictNullChecks` is indeed `false` in that case.

Since this is related to the same file being compiled multiple times, it may as well be related to this issue.

**Comment by higherorderfunctor** [comment]
To expand upon @fluidsonic's findings, using the file watchers (without actually watching the filesystem) looks like it may be a viable solution.

I created a small proof-of-concept to try and isolate certain behaviors.  It will unfortunately require some state management.

I track watches requested by tsserver in a Trie data structure to find the most specific watch.  Triggering it after creating a new file which fixed the inferred project issue.  This means it will use the correct `tsconfig.json` and not the default compiler options which is why the `strictNullChecks` issue pops up.

I'm sure more is required to integrate into the plugin efficiently since the plugin is just getting files from eslint without necessarily having access to if the file is new or changed since the last project update.

It doesn't really come with much performance benefits.  It was slower on my test single file repo, hanging while waiting for the `projectsUpdatedInBackground` event, even though the results were available almost a full second earlier in the logs.  It was slightly faster (almost by a second) on a large repo.

For `--fix` I'm still wondering if using something like [ScriptInfo.editContent](https://github.com/microsoft/TypeScript/blob/d8086f14b6b97c0df34a0cc2f56d4b5926a0c299/src/server/scriptInfo.ts#L678) would be faster.  That likely involves even more state management though.


<details>

<summary>Expand <b>src/index.ts</b>...</summary>

```ts
/* eslint-disable @typescript-eslint/no-non-null-assertion */
/* eslint-disable @typescript-eslint/no-empty-function */
import ts from 'typescript';
import path from 'node:path';
import fs from 'node:fs';

const eventHandler: ts.server.ProjectServiceEventHandler = (event) => {
  console.log(event);
};

const logger: ts.server.Logger = {
  close: () => {},
  endGroup: () => {},
  getLogFileName: (): undefined => undefined,
  hasLevel: () => true,
  info: (s) => {
    console.log('INFO:', s);
  },
  loggingEnabled: () => true,
  msg: (s) => {
    console.log('MSG:', s);
  },
  perftrc: (s) => {
    console.log('PERF:', s);
  },
  startGroup: () => {},
};

class TrieNode<T> {
  children: Map<string, TrieNode<T>>;
  value: T | null;
  constructor(public readonly path: string) {
    this.children = new Map();
    this.value = null;
  }
}

class Trie<T> {
  root: TrieNode<T>;

  constructor() {
    this.root = new TrieNode('');
  }

  insert(path: string) {
    if (!path.startsWith('/')) throw new Error('absolute paths only');
    const parts = path.split('/').slice(1); // drop the first empty string
    const { currentNode } = parts.reduce(
      ({ currentNode, rootPath }, part) => {
        path = `${rootPath}/${part}`;
        if (!currentNode.children.has(part)) {
          currentNode.children.set(part, new TrieNode(path));
        }
        return {
          currentNode: currentNode.children.get(part)!,
          rootPath: path,
        };
      },
      {
        currentNode: this.root,
        rootPath: this.root.path,
      },
    );
    return currentNode;
  }

  get(path: string): TrieNode<T> | null {
    const parts = path.split('/');
    const { lastNodeWithValue } = parts.reduce(
      ({ currentNode, lastNodeWithValue }, part) => {
        if (!currentNode.children.has(part)) {
          return { currentNode: currentNode, lastNodeWithValue };
        }
        const childNode = currentNode.children.get(part)!;
        return {
          currentNode: childNode,
          lastNodeWithValue: childNode.value !== null ? childNode : lastNodeWithValue,
        };
      },
      {
        currentNode: this.root,
        lastNodeWithValue: null as TrieNode<T> | null,
      },
    );
    return lastNodeWithValue;
  }
}

class Watcher implements ts.FileWatcher {
  constructor(
    private readonly node: TrieNode<Watcher>,
    public readonly callback: () => void,
  ) {}

  close() {
    this.node.value = null;
  }
}

const watchers = new Trie<Watcher>();

const stubFileWatcher = (
  path: string,
  callback: ts.FileWatcherCallback,
  pollingInterval?: number,
  options?: ts.WatchOptions,
): ts.FileWatcher => {
  console.log('WATCHFILE:', path, pollingInterval, options);
  const node = watchers.insert(path);
  if (node.value !== null) {
    return node.value;
  }
  const watcher = new Watcher(node, () => {
    callback(path, ts.FileWatcherEventKind.Changed, new Date());
  });
  node.value = watcher;
  return watcher;
};

const stubDirectoryWatcher = (
  path: string,
  callback: ts.DirectoryWatcherCallback,
  recursive?: boolean,
  options?: ts.WatchOptions,
): ts.FileWatcher => {
  console.log('WATCHDIR:', path, recursive, options);
  const node = watchers.insert(path);
  if (node.value !== null) {
    return node.value;
  }
  const watcher = new Watcher(node, () => {
    callback(path);
  });
  node.value = watcher;
  return watcher;
};

const host: ts.server.ServerHost = {
  ...ts.sys,
  clearImmediate,
  clearTimeout,
  setImmediate,
  setTimeout,
  watchFile: stubFileWatcher,
  watchDirectory: stubDirectoryWatcher,
};

const options: ts.server.ProjectServiceOptions = {
  host, // host capabilities
  logger, // logger
  cancellationToken: ts.server.nullCancellationToken, // cancel in flight requests
  useSingleInferredProject: false, // treat files without a tsconfig as separate projects
  useInferredProjectPerProjectRoot: false, // don't lump files without a tsconfig that are in the same directory together
  // typingsInstaller: undefined, // use default
  eventHandler, // get those diags
  canUseWatchEvents: true, // use setup watches
  suppressDiagnosticEvents: false, // we want the diagnostics
  throttleWaitMilliseconds: 0, // 200, // debounce requests
  globalPlugins: [], // no extra global plugins
  pluginProbeLocations: [], // just look for plugins in node_modules
  allowLocalPluginLoads: true, // allow plugins
  // typesMapLocation: undefined, // use default
  serverMode: ts.LanguageServiceMode.Semantic, // full type checking
  session: undefined, // lsp interface, this PoC has eslint as interface so don't use
  jsDocParsingMode: ts.JSDocParsingMode.ParseAll, // parse all jsdoc
};

const projectService = new ts.server.ProjectService(options);

const file = ts.server.toNormalizedPath(`${process.cwd()}/src/index.ts`);

const formatDiagnostic = (diagnostic: ts.Diagnostic) => {
  if (diagnostic.file === undefined) return; // TODO: could be whole project
  if (diagnostic.start === undefined) return; // TODO: could be whole file
  const { line, character } = diagnostic.file.getLineAndCharacterOfPosition(diagnostic.start);
  const message = ts.flattenDiagnosticMessageText(diagnostic.messageText, '\n');
  // eslint-disable-next-line @typescript-eslint/restrict-template-expressions
  console.log(`${diagnostic.file.fileName} (${line + 1},${character + 1}): ${message}`);
};

const printDiagnostic = (file: ts.server.NormalizedPath) => {
  console.log('GETTING DIAGNOSTICS FOR:', file);

  projectService.openClientFile(file, undefined, ts.ScriptKind.TS, process.cwd());

  const scriptInfo = projectService.getScriptInfo(file);

  if (!scriptInfo) {
    console.log('NO SCRIPT INFO');
    process.exit(1);
  }

  const project = projectService.getDefaultProjectForFile(
    scriptInfo.fileName,
    true, // finds project or creates inferred project
  );

  if (project) {
    const languageService = project.getLanguageService();

    const program = languageService.getProgram();

    if (program) {
      const source = program.getSourceFile(file);
      const diagnostics = program.getSemanticDiagnostics(source);
      diagnostics.forEach(formatDiagnostic);
    }
  }
};

printDiagnostic(file);

const destination = ts.server.toNormalizedPath(path.join(path.dirname(file), 'index2.ts'));

fs.copyFileSync(file, destination);

const watcher = watchers.get(destination);
if (watcher !== null && watcher.value !== null) watcher.value.callback();

printDiagnostic(destination);

fs.rmSync(destination);
```

</details>

With this block executed after creating a new file, tsserver finds and associates to the correct project.

```ts
const watcher = watchers.get(destination);
if (watcher !== null && watcher.value !== null) watcher.value.callback();
```

```
INFO:   FileName: <REDACTED>/ts-editor-services-test/src/index.ts ProjectRootPath: <REDACTED>/ts-editor-services-test
INFO:           Projects: <REDACTED>/ts-editor-services-test/tsconfig.json
INFO:   FileName: <REDACTED>/ts-editor-services-test/src/index2.ts ProjectRootPath: <REDACTED>/ts-editor-services-test
INFO:           Projects: <REDACTED>/ts-editor-services-test/tsconfig.json
```

<details>

<summary>More log output...</summary>

```
INFO: -----------------------------------------------
INFO: Search path: <REDACTED>/ts-editor-services-test
INFO: For info: <REDACTED>/ts-editor-services-test/tsconfig.json :: No config files found.
INFO: Project '<REDACTED>/ts-editor-services-test/tsconfig.json' (Configured)
INFO:   Files (173)

INFO: -----------------------------------------------
INFO: Open files:
INFO:   FileName: <REDACTED>/ts-editor-services-test/src/index.ts ProjectRootPath: <REDACTED>/ts-editor-services-test
INFO:           Projects: <REDACTED>/ts-editor-services-test/tsconfig.json
INFO:   FileName: <REDACTED>/ts-editor-services-test/src/index2.ts ProjectRootPath: <REDACTED>/ts-editor-services-test
INFO:           Projects: <REDACTED>/ts-editor-services-test/tsconfig.json
INFO: Running: <REDACTED>/ts-editor-services-test/tsconfig.json
INFO: Running: *ensureProjectForOpenFiles*
INFO: Before ensureProjectForOpenFiles:
INFO: Project '<REDACTED>/ts-editor-services-test/tsconfig.json' (Configured)
INFO:   Files (173)

INFO: -----------------------------------------------
INFO: Open files:
INFO:   FileName: <REDACTED>/ts-editor-services-test/src/index.ts ProjectRootPath: <REDACTED>/ts-editor-services-test
INFO:           Projects: <REDACTED>/ts-editor-services-test/tsconfig.json
INFO:   FileName: <REDACTED>/ts-editor-services-test/src/index2.ts ProjectRootPath: <REDACTED>/ts-editor-services-test
INFO:           Projects: <REDACTED>/ts-editor-services-test/tsconfig.json
INFO: After ensureProjectForOpenFiles:
INFO: Project '<REDACTED>/ts-editor-services-test/tsconfig.json' (Configured)
INFO:   Files (173)

INFO: -----------------------------------------------
INFO: Open files:
INFO:   FileName: <REDACTED>/ts-editor-services-test/src/index.ts ProjectRootPath: <REDACTED>/ts-editor-services-test
INFO:           Projects: <REDACTED>/ts-editor-services-test/tsconfig.json
INFO:   FileName: <REDACTED>/ts-editor-services-test/src/index2.ts ProjectRootPath: <REDACTED>/ts-editor-services-test
INFO:           Projects: <REDACTED>/ts-editor-services-test/tsconfig.json
{
  eventName: "projectsUpdatedInBackground",
  data: {
    openFiles: [ "<REDACTED>/ts-editor-services-test/src/index.ts",
      "<REDACTED>/ts-editor-services-test/src/index2.ts"
    ],
  },
}
```

</details>

If I comment that section out, the new file is assigned to an inferred project.

```
INFO:   FileName: <REDACTED>/ts-editor-services-test/src/index.ts ProjectRootPath: <REDACTED>/ts-editor-services-test
INFO:           Projects: <REDACTED>/ts-editor-services-test/tsconfig.json
INFO:   FileName: <REDACTED>/ts-editor-services-test/src/index2.ts ProjectRootPath: <REDACTED>/ts-editor-services-test
INFO:           Projects: /dev/null/inferredProject1*
```

<details>

<summary>More log output...</summary>

```
INFO: -----------------------------------------------
INFO: Project '<REDACTED>/ts-editor-services-test/tsconfig.json' (Configured)
INFO:   Files (172)

INFO: -----------------------------------------------
INFO: Project '/dev/null/inferredProject1*' (Inferred)
INFO:   Files (178)

INFO: -----------------------------------------------
INFO: Open files:
INFO:   FileName: <REDACTED>/ts-editor-services-test/src/index.ts ProjectRootPath: <REDACTED>/ts-editor-services-test
INFO:           Projects: <REDACTED>/ts-editor-services-test/tsconfig.json
INFO:   FileName: <REDACTED>/ts-editor-services-test/src/index2.ts ProjectRootPath: <REDACTED>/ts-editor-services-test
INFO:           Projects: /dev/null/inferredProject1*
```

</details>

**Comment by higherorderfunctor** [comment]
So I have a hobbled together [set of what I hope are the start of fixes](https://github.com/higherorderfunctor/typescript-eslint/tree/testing) .  One note if anyone clones, the `package.json`s were modified to use links for in-repo dependencies so I could add the build outputs to my larger work repository also using links but with pnpm.

First, I track watches set by the `ProjectService` with a new [getWatchesForProjectService.ts](https://github.com/higherorderfunctor/typescript-eslint/blob/testing/packages/typescript-estree/src/create-program/getWatchesForProjectService.ts).  Will need to look through [getWatchProgramsForProjects.ts](https://github.com/higherorderfunctor/typescript-eslint/blob/testing/packages/typescript-estree/src/create-program/getWatchProgramsForProjects.ts) to see if there is room to consolidate.

This is used by [createProjectService.ts](https://github.com/higherorderfunctor/typescript-eslint/blob/testing/packages/typescript-estree/src/create-program/createProjectService.ts) to call the watch constructors via the host and [useProgramFromProjectService.ts](https://github.com/higherorderfunctor/typescript-eslint/blob/testing/packages/typescript-estree/src/useProgramFromProjectService.ts) to trigger the watches when a file has not been seen by the `ProjectService`.  I trigger the most specific watch which fixed the `strictNullChecks` for me.

Also noteworthy is that calling `service.setHostConfiguration` triggers a full project reload for files with irregular extensions like `vue`.  It should probably be set once with `ProjectService` during construction instead of on every file.  Linting a vue package in my monorepo went from 3m30s to 20s.  The whole monorepo I am testing on went from 16m in CI on server hardware to 3m40s on the mini pc I am currently using, which is a much slower machine.  I'm excited by those numbers and they produced the same results for both static analysis and running with `--fix`.

I may have a memory leak as I had to increase the heap.  Could also just be linting a large monorepo using only a default project with`ProjectService`s without ever closing a file with tsserver.

**Comment by higherorderfunctor** [comment]
If anyone wants to test, I've moved to this branch: https://github.com/higherorderfunctor/typescript-eslint/tree/patches2upstream.  There are a lot of modifications, so I'll work to get these up-streamed over several PRs.  I've confirmed the `strictNullChecks` issue is still fixed for the scenario I was able to reproduce.  I'm still on `eslint@8.57.0` for reference as there is one plugin I use that doesn't work with `FlatCompat` in `eslint@9` (`eslint-config-airbnb-base/whitespace`).

I've included the `types` and `typescript-estree` builds in the repo if your package manager supports installing from git.  My `package.json` looks like below with `pnpm@9.1.x`.  The patch is for https://github.com/typescript-eslint/typescript-eslint/issues/9223 until those fixes make it into [eslint-plugin-import-x](https://github.com/antfu/eslint-plugin-import-x/blob/4dac1cb2ef94e6357f4ae84185358f69f626cfb7/src/utils/parse.ts#L92).

<details>

<summary>package.json</summary>

```json
{
  ...
  "pnpm": {
    "patchedDependencies": {
      "eslint-plugin-import-x@0.5.1": "patches/eslint-plugin-import-x@0.5.1.patch",
    },
    // forces dependencies to use the same version
    "overrides": {
      "typescript": "$typescript",
      "typescript-eslint": "$typescript-eslint",
      "@typescript-eslint/eslint-plugin": "$typescript-eslint",
      "@typescript-eslint/parser": "$typescript-eslint",
      "@typescript-eslint/rule-schema-to-typescript-types": "$typescript-eslint",
      "@typescript-eslint/scope-manager": "$typescript-eslint",
      "@typescript-eslint/type-utils": "$typescript-eslint",
      "@typescript-eslint/utils": "$typescript-eslint",
      "@typescript-eslint/visitor-keys": "$typescript-eslint",
      // my custom fork
      "@typescript-eslint/types": "higherorderfunctor/typescript-eslint#patches2upstream&path:packages/types",
      "@typescript-eslint/typescript-estree": "higherorderfunctor/typescript-eslint#patches2upstream&path:packages/typescript-estree"
    }

  },
  "devDependencies": {
    "eslint": "^8.57.0",
    "typescript": "rc",
    "typescript-eslint": "rc-v8",
    // just for reference on what I have tested against
    "@eslint/eslintrc": "^3.1.0",
    "@eslint/js": "^8.57.0",
    "@stylistic/eslint-plugin": "^2.1.0",
    "@types/eslint": "^8.56.10",
    "@typescript-eslint/utils": "rc-v8", // uses some types in my eslint.config.js for type checking
    "@vue/eslint-config-prettier": "^9.0.0",
    "@vue/eslint-config-typescript": "^13.0.0",
    "eslint-config-airbnb-base": "^15.0.0",
    "eslint-config-airbnb-typescript": "^18.0.0",
    "eslint-config-prettier": "^9.1.0",
    "eslint-import-resolver-alias": "^1.1.2",
    "eslint-import-resolver-typescript": "^3.6.1",
    "eslint-plugin-codegen": "0.28.0",
    "eslint-plugin-deprecation": "^3.0.0",
    "eslint-plugin-import-x": "^0.5.1",
    "eslint-plugin-jsonc": "^2.16.0",
    "eslint-plugin-markdown": "^5.0.0",
    "eslint-plugin-prefer-arrow-functions": "^3.3.2",
    "eslint-plugin-prettier": "^5.1.3",
    "eslint-plugin-promise": "^6.2.0",
    "eslint-plugin-security": "^3.0.0",
    "eslint-plugin-simple-import-sort": "^12.1.0",
    "eslint-plugin-sonar": "^0.14.1",
    "eslint-plugin-sonarjs": "^1.0.3",
    "eslint-plugin-sort-destructure-keys": "^2.0.0",
    "eslint-plugin-tsdoc": "^0.3.0",
    "eslint-plugin-vue": "^9.26.0",
    "eslint-plugin-vuejs-accessibility": "^2.3.0",
    "eslint-plugin-vuetify": "^2.4.0",
    "eslint_d": "^13.1.2",
    "jsonc-eslint-parser": "^2.4.0",
    "vue-eslint-parser": "^9.4.3",
    ...
  },
  ...
}
```

</details>

My RAM utilization is way down today... not sure if it is something I did, something upstream in the `rc-v8` branch, or something in the latest `typescript@rc`, but linting my whole project now is down to ~2min and RAM is slightly over 3G.  I had my heap up to 9G a couple days ago or it would crash.

```
        Elapsed (wall clock) time (h:mm:ss or m:ss): 2:04.05
        Maximum resident set size (kbytes): 3235328
```

If you do get OOM, the heap can be increased.

```sh
NODE_OPTIONS="--max-old-space-size=8184" eslint ...
```

**Known Issues**

With the flat config and `extraFileExtensions`, it is best to set it once if possible.  I had different `projectService` settings between my front-end and back-end.  Changing the `extraFileExtensions` mid-lint causes tsserver to do a full project reload.  I have guards in place to do it, but it will cause a performance hit.  I originally only set it once and was wondering why I was having issue with `.vue` files intermittently.  Was caused by which `projectService` happened to be loaded first.

**Additional Options**

I've added two new options with their defaults below.  `incremental` only sends deltas to tsserver which I am hoping makes the LSP experience better.  `maximumOpenFiles` will automatically close files with tsserver, which is nice on the CLI which has no mechanism to know when it is done with a file.  The new options are why the `types` package is needed if you enable typing on your `eslint.config.js`.

```js
{
  languageOptions: {
    parserOptions: {
      projectService: {
        incremental: false, // only sends changes to tsserver instead of whole files
        maximumOpenFiles: 16, // auto closes open files with tsserver based on access order
      }
    }
  }
}
```

**Comment by JoshuaKGoldberg** [comment]
Marking as accepting PRs since I'm conferencing much of this month. Caveats:
* That this is very TypeScript-APIs-heavy stuff. Prepare for deep diving into TS source.
* This is time sensitive for us, so we might have to take over a PR that's stalled.

**Comment by higherorderfunctor** [comment]
I mentioned this in another thread, but I have a busy week this week in case someone else needs to pick up the PR.  Dropping my notes in case anyone else does pick it up.

The working implementation is https://github.com/higherorderfunctor/typescript-eslint/tree/patches2upstream, the draft PR is not currently in a working state.

While I have a working implementation, it duplicates some of the existing logic in `getWatchProgramsForProjectsts.ts`.

`saveWatchCallback` looks likely to be a drop in candidate for my implementation.

Exporting or proving an access function for `fileWatchCallbackTrackingMap` and `folderWatchCallbackTrackingMap` could allow for the watches to be updated easily.  I haven't analyzed the paths tsserver requests the host to watch.  I opted to use a trie so I could find the most specific path.  `getWatchProgramsForProjectsts.ts` just uses a `Map`, so maybe supporting partial matches is not even needed.

`hasTSConfigChanged` is probably good to consider since changes to `tsconfig.json` I don't believe will reported to the plugin.

`maybeInvalidateProgram` which calls the above has quite a bit of logic, but I am not sure all of it is needed for project services.  I didn't see a need to do all the caching that is done.  Linting results were almost immediate by just walking the service, even on a larger project, and trigger the watcher for the most specific matching path.

```ts
const filePathMatchedByConfiguredProject = (
  service: ts.server.ProjectService,
  filePath: string,
): boolean => {
  const configuredProjects = service.configuredProjects;
  for (const project of configuredProjects.values()) {
    if (project.containsFile(filePath as ts.server.NormalizedPath)) {
      return true;
    }
  }
  return false;
};
```

**Comment by JoshuaKGoldberg** [comment]
👋 @higherorderfunctor now that #9306 is merged _(thanks again!)_, do you think you'll have time to un-draft #9353 in the next few days?

**Comment by JoshuaKGoldberg** [comment]
👋 @higherorderfunctor now that #9306 is merged _(thanks again!)_, do you think you'll have time to un-draft #9353 in the next few days? For context, we plan on releasing v8 in three days: midday Wednesday EST.

I'm also wondering if this is just not something we can reasonably be tackling internally in typescript-eslint. ESLint's architecture right now doesn't give us any indication as to whether it's single-run or when files are closed in an editor (https://github.com/eslint/rfcs/pull/102). We have existing issues with file changes in editors (see #3536 & its links)...

**Comment by higherorderfunctor** [comment]
@JoshuaKGoldberg I can probably get a minimum viable implementation done by Monday.  Basically, is the file in the project service? If no, trigger the best matching watcher.  Recheck and if still no add to default project or treat as an isolated file.

This will deal with new source files requiring the LSP to be restarted.

Some of the more complex decision making I don't think I'd be able to have ready by release day.  Examples:

- Handling changes to tsconfig.
- Handling changes to package.json and/or node_modules to reload dependencies.
- Handling changes to deleted or moved files.  This shouldn't impact runtime behavior, but the indexing of watchers would hold refs to non existent files and directories leaking some memory.
- Looking for potential optimizations.  Basic one I could implement is not indexing or tracking file watchers.  File changes are handled by opening the files and sending "editor" updates and not through watchers.  That one is easy, but there may be others I haven't considered.
- Is this even the best approach long term?  Users will most likely be using the tsserver lsp which is doing real watching including tsconfigs and dependencies.  Sharing the same project service instance could solve these project structure updates.  A lot to unpack with that idea though.

**Comment by JoshuaKGoldberg** [comment]
Yeah, these are a lot of open questions. I really _really_ appreciate your offer to hustle but long-term I don't think it's a given that we'd want to try to handle these cases internally (https://github.com/typescript-eslint/typescript-eslint/issues/8835#issuecomment-2254489382). I'm going to go ahead and remove this from the v8 milestone. 

> Sharing the same project service instance could solve these project structure updates

That's one of the things I'm very hopeful for with the project service!

**Comment by leaftail1880** [comment]
Also experiencing this issue with `typescript-eslint@8.0.1` and `strictTypeChecked`/`stylisticTypeChecked` configs in **monorepo**. Each time i create a new file, it errors `'this rule requires strictNullChecks in tsconfig.json'` and simply reloading ESLint server using VSCode command helps, but its just annoying.

strict is actually **enabled** on each config in monorepo

My eslint.config.js:
```js
import eslint from '@eslint/js'
import prettier from 'eslint-config-prettier'
import ts from 'typescript-eslint'

export default ts.config(
  eslint.configs.recommended,
  prettier,
  ...ts.configs.stylisticTypeChecked,
  ...ts.configs.strictTypeChecked,
  {
    languageOptions: { parserOptions: { projectService: true, tsconfigRootDir: import.meta.dirname }
  }
)
```

| Package | Version |
|--------|--------|
| node | v22.1.4 |
| typescript | v5.5.4 |
| typescript-eslint | v8.0.1 | 
| eslint | v9.6.0 |


UPDATE: Never experienced this again after updating to `typescript-eslint@8.35.0`, seems fixed to me

**Comment by JoshuaKGoldberg** [comment]
It's been over a year since we added the `evaluating community engagement` label. In that time:

* Exactly one comment has been posted
* We've repeatedly butted against limitations of ESLint's architecture trying to fix similar issues: #11677

I'm going to take that as a signal that this issue is not something we can commit to. We'd _like_ to solve it but per #11677 it's _very_ difficult to work with the current ESLint parsing architecture. If someone figures out a good way to solve for _"multiple tsconfigs in persistent linting states"_ please do post here or in a new issue/PR and we can re-open. But until then we'll have to close this out.

Thanks for the discussions & reproductions all! 💔
